docs/csharp/local-functions-vs-lambdas.md

```diff
@@ -26,46 +27,87 @@ Contrast that implementation with a version that uses lambda expressions:
+The local functions have names. The lambda expressions are anonymous methods
+that are assigned to variables that are `Func` or `Action` types. When you
+declare a local function the argument types and return type are part of the 
+function declaration. The argument types and return type are part of the
+variable type declaration when you declare a lambda expression instead of
+being part of the body of the lambda expression. Those two difference may
+result in more clear code.
+
+Second, local functions have different rules for definite assignment
+than lambda expressions. A local function declaration can be referenced
+from any code location where it is in scope. A lambda expression must be
+assigned before it can be accessed (or called through a delgate
+referencing the lambda expression.) Notice that the version using the
+lambda expression must declare and initialize the lambda expression,
+`nthFactorial` before defining it. Not doing so results in a compile
+time error for referencing `nthFactorial` before assigning it.
 Recursive algorithms are easier to create using local functions.
 
-Third, for lambda expressions, the compiler must always create an anonymous class
-and an instance of that class to store any variables captured by the
-closure. Consider this async example:
+Definite assignment rules also affect any variables that are captured
+by the local function or lamdba epression. Both local functions and
+lambda expression rules demand that any captured variables are definitely
+assigned at the point when the local function or lambda expression is
+converted to a delegate. The difference is that lambda expressions are converted
+to delegates when they are declared. Local functions are converted to delegates
+only when used as a delegate. If you declare a local function, and only
+reference it by calling it like a method, it will not be converted to
+a delegate. That rule enables you to declare
+a local function at any convenient location in its enclosing scope. It's common
+to declare local functions at the end of the parent method, after any return
+statements.
+
+Third, the compiler can perform static analysis that enables local functions to
+definitely assign captured variables in the enclosing scope.
+
+Consider this example:
+
+```csharp
+bool M()
+{
+    int y;
+    Local();
+    return y;
+
+    void Local() => y = 0;
+}
+```
+
+The compiler can determine that `Local` definitely assigns `y` when called. Because `Local` is called before the `return` statement, `y` is definitiely
+assigned at the `return` statement.
+
+The analysis that enables that analysis enables the fourth difference.
+Depending on their use, local functions can avoid heap allocations that
+are always necessary for lambda expressions. If a local function is never
+converted to a delegate and none of the variables captured by a locatl function
+are converted to delegates, the compiler can avoid heap allocations. 
+
+Consider this async example:
```

Comment:
nit: there's no need for a paragraph break here.

Reply:
Thanks for the review @rpetrusha I've addressed all your comments.